core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.java

```diff
@@ -131,8 +131,26 @@ public void readAloud(WebView webView) {
           context.getResources().getString(R.string.tts_lang_not_supported),
           Toast.LENGTH_LONG).show();
       } else {
+        // check if voice needs to be installed for API below 21
+        if(VERSION.SDK_INT < VERSION_CODES.LOLLIPOP
+          && tts.getFeatures(locale).contains(TextToSpeech.Engine.KEY_FEATURE_NOT_INSTALLED)){
+          Toast.makeText(context,
+            context.getResources().getString(R.string.tts_lang_not_supported),
+            Toast.LENGTH_LONG).show();
+          return;
+        }
+
         tts.setLanguage(locale);
 
+        // check if voice needs to be installed for API 21 or higher
+        if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP
```

Comment:
these 2 new execution branches are very similar, what actually happens is you switch the source of the `features` depending on api level so I would write it so it does that

Reply:
@macgills They are exactly similar but the for API 21 or higher I have to execute `tts.setLanguage(locale)` in line 143 before searching for features to avoid using deprecated methods. However APIs before 21 did not have this requirement. I could combine them in one if condition but I will have to execute `tts.setLanguage(locale)` in both cases. I am not sure what is the best practice for something like this, do you think I should combine them and execute line 143 in both cases ?

Reply:
Yes, I don't really care if we set the locale first, we either play the audio or we don't.

Reply:
please review the last commit 2ead103